### PR TITLE
Refactor/kws 1888 update user

### PR DESCRIPTION
**Pull request by GuilhermeMota93**

### CODE REVIEW

File: KWSiOSSDKObjC/Classes/Base/KWSSDK.swift [hunk 1]
```diff
@@ -55,7 +55,7 @@ import SAMobileBase
         }
     }
     
-    public func  getKWSMetadata (token: String) -> MetadataKWS? {
+    public func getKWSMetadata (token: String) -> MetadataKWS? {
```
Comment (devgabrielcoman) [inline KWSiOSSDKObjC/Classes/Base/KWSSDK.swift]: hmm, why is this method here? Why should our **public facing SDK interface** have a method that creates `KWS metadata`?

File: KWSiOSSDKObjC/Classes/Base/KWSSDK.swift [hunk 2]
```diff
@@ -67,4 +67,31 @@ import SAMobileBase
         
         return metadata
     }
+    
+    public func getUserDetailsCountryCode(country: String) -> String?{
```
Comment (devgabrielcoman) [inline KWSiOSSDKObjC/Classes/Base/KWSSDK.swift]: Same for here:
- why is this method here?
- why **do** we even have this method? why do we need to map 

```
case "gb", "GB", "United Kingdom":
  return "gb"
```

File: KWSiOSSDKObjC/Classes/Base/Requests/UpdateUserDetailsRequest.swift [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        addBodyUserDetailsEmail()
+        addBodyUserDetailsGender()
+        addBodyUserDetailsLanguage()
+        addBodyUserDetailsAddressJSONString()
+        addBodyUserDetailsAppProfileJSONString()
+        
+    }
+    
+    private func addBodyUserDetailsFirstName(){
+        if (userDetails.firstName != nil && !(userDetails.firstName?.isEmpty)!){
```
Comment (devgabrielcoman) [inline KWSiOSSDKObjC/Classes/Base/Requests/UpdateUserDetailsRequest.swift]: Having:

```
(userDetails.firstName?.isEmpty)!
```

isn't it dangerous? Explicitly calling an optional value may result in crashes. How about doing it like this:

```
if let firstName = userDetails.firstName, !firstName.isEmpty {
  self.body?["firstName"] = firstName
}
```

File: KWSiOSSDKObjC/Classes/Base/Requests/UpdateUserDetailsRequest.swift [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    
+    private func addBodyUserDetailsLanguage(){
+        if (userDetails.language != nil && !(userDetails.language?.isEmpty)!){
+            self.body!["language"] = userDetails.language
+        }
+    }
+    
+    private func addBodyUserDetailsAddressJSONString(){
+        
+        if(userDetails.address != nil){
```
Comment (devgabrielcoman) [inline KWSiOSSDKObjC/Classes/Base/Requests/UpdateUserDetailsRequest.swift]: Again,

```
if let address = userDetails.address {
 // ...
}
```

Comment (devgabrielcoman) [inline KWSiOSSDKObjC/Classes/Base/Requests/UpdateUserDetailsRequest.swift]: Hmm, this whole block can be expressed better, in a more swift / logical way:

```
let encoder = JSONEncoder () 

if let address = userDetails.address, 
   let jsonData = try? encoder.encode(address),
   let jsonString = String(data: jsonData, encoding: .utf8),
   !jsonString.isEmpty && jsonString != "{}"  {
   self.body?["address"] = jsonString
}
```

so basically:
1) we've removed all **null**-ness checks (very java / c-style code, not so much swift / kotlin / optional-based)
2) we have less code and less paths to go on

Comment (GuilhermeMota93) [reply KWSiOSSDKObjC/Classes/Base/Requests/UpdateUserDetailsRequest.swift]: Agree, will update it 👌

File: KWSiOSSDKObjC/Classes/Base/Requests/UpdateUserDetailsRequest.swift [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            
+            if(!userDetailsJSONString.isEmpty && userDetailsJSONString != "{}"){
+                self.body!["address"] = userDetailsJSONString
+            }
+        }
+    }
+    
+    private func addBodyUserDetailsAppProfileJSONString(){
+        
+        if(userDetails.applicationProfile != nil){
```
[the remark this answers is not in the document]
Comment (GuilhermeMota93) [reply KWSiOSSDKObjC/Classes/Base/Requests/UpdateUserDetailsRequest.swift]: Will update as per comment above 👌

File: KWSiOSSDKObjC/Classes/Base/Requests/UpdateUserDetailsRequest.swift (hunk 3 above)
Comment (GuilhermeMota93) [reply KWSiOSSDKObjC/Classes/Base/Requests/UpdateUserDetailsRequest.swift]: Looks cleaner, will update!

File: KWSiOSSDKObjC/Classes/Base/Requests/UpdateUserDetailsRequest.swift [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            var userDetailsAppProfileJSONString = ""
+            
+            if(jsonDataUserDetailsAppProfile != nil){
+                userDetailsAppProfileJSONString = String(data: jsonDataUserDetailsAppProfile!, encoding: .utf8)!
+                print("User Details App Profile as JSON string -> \n" + userDetailsAppProfileJSONString + "\n")
+            }else{
+                print("Couldn't parse the user details address...something went wrong!")
+            }
+            
+            if(!userDetailsAppProfileJSONString.isEmpty && userDetailsAppProfileJSONString != "{}"){
```
Comment (devgabrielcoman) [inline KWSiOSSDKObjC/Classes/Base/Requests/UpdateUserDetailsRequest.swift]: Alternatively, as a final note on this class, this whole mess of methods, etc., can be expressed more elegantly like this:

```
public class UpdateUserDetailsRequest: BaseRequest {
  private let userDetails: UserDetails

  public init(environment: KWSNetworkEnvironment,
                    userDetails: UserDetails,
                    userId: Int,
                    token: String) {
          // copy user details
          self.userDetails = userDetails
          // call super
          super.init(environment: environment, token: token)
 } 

 public var endpoint: String = "v1/users/\(userId)"
 public var method: HTTPMethod = .PUT 

 public var body: [String: Any]? {
   get {
      var tmpBody: [String:Any] = [:]
      
      // add all of the things that were previously in different methods
      if let firstName = userDetails.firstName, !firstName.isEmpty {
        self.body?["firstName"] = firstName
      }
      // even more stuff from methods ...
      return tmpBody
   }
 }
}
```

File: KWSiOSSDKObjC/Classes/Base/Responses/UserDetails/Points.swift [hunk 7]
```diff
@@ -32,6 +32,16 @@ public final class Points: NSObject {
 
     }
     
+    public enum CodingKeys: String, CodingKey {
```
Comment (devgabrielcoman) [inline KWSiOSSDKObjC/Classes/Base/Responses/UserDetails/Points.swift]: what's `CodingKeys` ?

File: KWSiOSSDKObjC/Classes/Base/Responses/UserDetails/UserAddress+Mapping.swift [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    }
+}
+
+
+extension UserAddress: Encodable {
+    
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: UserAddress.CodingKeys.self)
+        
+        if(street != nil && !(street?.isEmpty)!){
```
Comment (devgabrielcoman) [inline KWSiOSSDKObjC/Classes/Base/Responses/UserDetails/UserAddress+Mapping.swift]: I've seen this style of encoding in other models? Is this the new Swift 4 encoding? Why is it "so ugly"? 

Also, every time you do this `(street?.isEmpty)!` you run the very real risk of an **app crash** because in Swift, just like in Kotlin, forcibly unwrapping an optional will cause an crash (similar to `NullPointerException` in Java)

what you should do is:

```
if let mStreet = street, !mStreet.isEmpty {
 // bla bla 
}
```

Check this for some guidance: https://www.andrewcbancroft.com/2017/04/19/how-to-unwrap-multiple-optionals-with-one-if-let-in-swift/

Other than that, Swift allows you to do this:

```
if [condition_1], [condition_2], [condition_3_that_might_depend_on_1_and_2] { 
  // do something
}
```

So basically you can do, as an example:

```
if let myValue1 = someOptional1, let myValue2 = someOptional2, myValue1.name == myValue2.name {
  printf("Found some users!")
}
```

File: KWSiOSSDKObjC/Classes/Base/Responses/UserDetails/UserDetails.swift [hunk 9]
```diff
@@ -51,17 +56,39 @@ public final class UserDetailsResponse: NSObject {
...
         self.hasSetParentEmail = hasSetParentEmail
         self.applicationProfile = applicationProfile
         self.applicationPermissions = applicationPermissions
         self.points = points
         self.createdAt = createdAt
     }
     
+    public enum CodingKeys: String, CodingKey {
+        
+        //to encode
```
Comment (devgabrielcoman) [inline KWSiOSSDKObjC/Classes/Base/Responses/UserDetails/UserDetails.swift]: Also, why do some objects have `CodingKeys` and some objects have encoding methods?

File: KWSiOSSDKObjC/Classes/Base/KWSSDK.swift (hunk 2 above)
Comment (GuilhermeMota93) [reply KWSiOSSDKObjC/Classes/Base/KWSSDK.swift]: This was raised with Connor:
- https://jira.superawesome.tv/browse/KWS-1944
- https://jira.superawesome.tv/browse/KWS-1943

So basically, the *GET* user details returns a `country` as plain text (e.g "United Kingdom").

For the update user logic (*PUT request*), there's a validation (*in the app side, not the SDK*) if a previous user was retrieved (*GET request*) and, if so, add the changes/updates to the retrieved model and send it.  

While doing this logic, retrieving the `country` from the *GET* user (e.g "United Kingdom") and building the new body to send, the API does not complain about it and gives a 204 success. Although this is a success response, it's a wrong one. The API only uses (per their on mapping) a country code. Therefor, trying to do a *GetUserDetails* after *UpdateUserDetail* gives back an error (see JIRA items).

As changing this now might cause breaking changes in live users, the short term fix was to map on the app side.

Suggestions for the future:
- have the API return an extra `countryCode` in the *GET response*. We can then use this to set the `country` when doing a *PUT request*, instead of the mapping;
- improve API PUT handling to not allow anything it's not mapped on their side and give a `Only country codes allowed for country` error

File: KWSiOSSDKObjC/Classes/Base/KWSSDK.swift (hunk 1 above)
Comment (GuilhermeMota93) [reply KWSiOSSDKObjC/Classes/Base/KWSSDK.swift]: Not the best place to have it, I agree, but due to the need to parse the Metadata using SAMobileBase and to avoid more complications on Objective-C/Swift, as `ParseBase64Request` and `ParseBase64Task` can't be accessed in the `KWSChildren.m` class the I opted to put this logic in the existing singleton.

I can create a Helper class for this kind of things, where I can have this method and the mapping country one. What's your thoughts on that?

File: KWSiOSSDKObjC/Classes/Base/Requests/UpdateUserDetailsRequest.swift (hunk 6 above)
Comment (GuilhermeMota93) [reply KWSiOSSDKObjC/Classes/Base/Requests/UpdateUserDetailsRequest.swift]: 👌

File: KWSiOSSDKObjC/Classes/Base/Responses/UserDetails/Points.swift (hunk 7 above)
Comment (GuilhermeMota93) [reply KWSiOSSDKObjC/Classes/Base/Responses/UserDetails/Points.swift]: `CodingKeys` can be used a little bit as the `@SerializableName` in Android from what I understood (see https://hackernoon.com/everything-about-codable-in-swift-4-97d0e18a2999 on the bit about "Choosing Properties to Encode and Decode — CodingKeys" ).

I tried to follow the PJ SDK 'recipe' for the Decoding bit, but due to the need to Encode in this case, I looked up multiple articles and this seemed the best approach. Might be a bit of boiler plate code as of now, but in case we want to use the same logic for both Decoding and Encoding, this can be the middle step (see snippet of code in the link above, just before "Limitations").

File: KWSiOSSDKObjC/Classes/Base/Responses/UserDetails/UserAddress+Mapping.swift (hunk 8 above)
Comment (GuilhermeMota93) [reply KWSiOSSDKObjC/Classes/Base/Responses/UserDetails/UserAddress+Mapping.swift]: I agree, it's not pretty as it is. Will have a look at your suggestion and try and make it nicer 👌

File: KWSiOSSDKObjC/Classes/Base/Responses/UserDetails/UserDetails.swift (hunk 9 above)
Comment (GuilhermeMota93) [reply KWSiOSSDKObjC/Classes/Base/Responses/UserDetails/UserDetails.swift]: Don't quite understand this comment. The encoding logic is in the extension of the object. I might of left something behind by mistake, but I'm not seeing it. Can you elaborate please?
